homeassistant/components/snapcast/media_player.py

```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
-    """Handle the entity service unjoin."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Entity is not a client. Can only unjoin clients.")
-    await entity.async_unjoin()
+            # Update known IDs
+            known_ids -= ids_to_remove
+            known_ids |= ids_to_add
 
+            return ids_to_add, ids_to_remove
 
-async def handle_set_latency(entity, service_call):
-    """Handle the entity service set_latency."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Latency can only be set for a Snapcast client.")
-    await entity.async_set_latency(service_call.data[ATTR_LATENCY])
+        new_entities: list[SnapcastBaseDevice] = []
 
+        group_ids = {g.identifier for g in coordinator.server.groups}
+        groups_to_add, groups_to_remove = _update_known_ids(_known_group_ids, group_ids)
 
-class SnapcastGroupDevice(MediaPlayerEntity):
-    """Representation of a Snapcast group device."""
+        new_entities.extend(
+            [
+                SnapcastGroupDevice(
+                    coordinator, coordinator.server.group(group_id), host_id
+                )
+                for group_id in groups_to_add
+            ]
+        )
+
+        client_ids = {c.identifier for c in coordinator.server.clients}
+        clients_to_add, clients_to_remove = _update_known_ids(
+            _known_client_ids, client_ids
+        )
+
+        new_entities.extend(
+            [
+                SnapcastClientDevice(
+                    coordinator, coordinator.server.client(client_id), host_id
+                )
+                for client_id in clients_to_add
+            ]
+        )
+
+        _LOGGER.debug(
+            "New clients: %s",
+            str([coordinator.server.client(c).friendly_name for c in clients_to_add]),
+        )
+        _LOGGER.debug(
+            "New groups: %s",
+            str([coordinator.server.group(g).friendly_name for g in groups_to_add]),
+        )
+        _LOGGER.debug(
+            "Remove client IDs: %s",
+            str([list(clients_to_remove)]),
+        )
+        _LOGGER.debug(
+            "Remove group IDs: %s",
+            str(list(groups_to_remove)),
+        )
+
+        # Add new entities
+        async_add_entities(new_entities)
+
+        # Remove stale entities
+        entity_registry = er.async_get(hass)
+        for group_id in groups_to_remove:
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastGroupDevice.get_unique_id(host_id, group_id),
+            ):
+                entity_registry.async_remove(entity_id)
+
+        for client_id in clients_to_remove:
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastClientDevice.get_unique_id(host_id, client_id),
```

Comment:
I think I prefer to have the `get_unique_id` functions to be outside of the class

Reply:
The unique_id format changes depending on if it's for a client or group, so not sure how I can move this out of the classes.

Reply:
Oh but its fine to have 2 functions, I just don't think we should use a static method for it

Reply:
Hm...I'm not sure I follow. If I made them instance methods I wouldn't be able to use them in this context. I'd have to construct temporary objects and that seems wasteful.

Reply:
No I mean, make them outside of the class. just have a `def _get_client_unique_id(.....)` and call that here and from the entity

Reply:
Ah gotcha.

Reply:
Ok, so a side effect of that change, I'd have to define constructors for both `SnapcastClientDevice` and `SnapcastGroupDevice` so they can call `_get_client_unique_id` and `_get_group_unique_id` respectively.

The classmethod approach allows the `SnapcastBaseDevice` constructor to handle everything.

It's not the end of the world but using the classmethod seems like the obvious choice, and I'm not sure what your concerns are with it.

Reply:
Oh you don't have a constructor now, hmmmm. Then maybe this works what you have now.